src/Interpreters/ClusterDiscovery.cpp

```diff
@@ -0,0 +1,369 @@
+        secure);
+    return cluster;
+}
+
+/// Reads data from zookeeper and tries to update cluster.
+/// Returns true on success (or no update required).
+bool ClusterDiscovery::updateCluster(ClusterInfo & cluster_info)
+{
+    LOG_TRACE(log, "Updating cluster '{}'", cluster_info.name);
+
+    auto zk = context->getZooKeeper();
+
+    int start_version;
+    Strings node_uuids = getNodeNames(zk, cluster_info.zk_root, cluster_info.name, &start_version, false);
+
+    if (node_uuids.empty())
+    {
+        LOG_ERROR(log, "Can't find any node in cluster '{}', will register again", cluster_info.name);
+        registerInZk(zk, cluster_info);
+        return false;
+    }
+
+    auto & nodes_info = cluster_info.nodes_info;
+    if (!needUpdate(node_uuids, nodes_info))
+    {
+        LOG_TRACE(log, "No update required for cluster '{}'", cluster_info.name);
+        return true;
+    }
+
+    nodes_info = getNodes(zk, cluster_info.zk_root, node_uuids);
+    if (nodes_info.empty())
+        return false;
+
+    int current_version;
+    getNodeNames(zk, cluster_info.zk_root, cluster_info.name, &current_version, true);
+
+    if (current_version != start_version)
+    {
+        nodes_info.clear();
+        return false;
+    }
+
+    auto cluster = getCluster(cluster_info);
+    context->setCluster(cluster_info.name, cluster);
+    return true;
+}
+
+bool ClusterDiscovery::updateCluster(const String & cluster_name)
+{
+    auto cluster_info = clusters_info.find(cluster_name);
+    if (cluster_info == clusters_info.end())
+    {
+        LOG_ERROR(log, "Unknown cluster '{}'", cluster_name);
+        return false;
+    }
+    return updateCluster(cluster_info->second);
+}
+
+void ClusterDiscovery::registerInZk(zkutil::ZooKeeperPtr & zk, ClusterInfo & info)
+{
+    String node_path = getShardsListPath(info.zk_root) / node_name;
+    zk->createAncestors(node_path);
+
+    NodeInfo self_node(getFQDNOrHostName() + ":" + toString(server_port));
+
+    zk->createOrUpdate(node_path, self_node.serialize(), zkutil::CreateMode::Ephemeral);
+    LOG_DEBUG(log, "Current node {} registered in cluster {}", node_name, info.name);
+}
+
+void ClusterDiscovery::start()
+{
+    if (clusters_info.empty())
+    {
+        LOG_DEBUG(log, "No defined clusters for discovery");
+        return;
+    }
+    LOG_TRACE(log, "Starting working thread");
+    main_thread = ThreadFromGlobalPool([this] { runMainThread(); });
+}
+
+void ClusterDiscovery::runMainThread()
+{
+    LOG_DEBUG(log, "Worker thread started");
+    // setThreadName("ClusterDiscovery");
+
+    {
+        auto zk = context->getZooKeeper();
+        for (auto & [_, info] : clusters_info)
+        {
+            registerInZk(zk, info);
+            if (!updateCluster(info))
+            {
+                LOG_WARNING(log, "Error on updating cluster '{}', will retry", info.name);
+                clusters_to_update->set(info.name);
+            }
+        }
+    }
```

Comment:
It looks like initial initialization should be done in the main thread, not in background thread, thus if server accepting connections then clusters will 100% available (without any possible races)

Reply:
There are pros and cons, at first I've done it like that, but then decided not to block the main thread for initial registration/update, because it communicates with zk and may take time. 
But, yes, initialization in `start` seems logic too, so, now I returned in into it.